janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java

```diff
@@ -1809,14 +1810,39 @@ private void testNestedWrites(String initialValue, String updatedValue, TestInfo
         propDeleter.commit();
 
         // The vertex must not exist after deletion
+        // See https://github.com/JanusGraph/janusgraph/issues/2176. The vertex is deleted from storage backend, but
+        // may not be deleted from index backend
         graph.tx().rollback();
         assertNull(getV(graph, id));
-        assertEmpty(graph.query().has(propName).vertices());
+        assertTrue(verticesRemoved(graph.query().has(propName).vertices()));
         if (null != updatedValue)
-            assertEmpty(graph.query().has(propName, updatedValue).vertices());
+            assertTrue(verticesRemoved(graph.query().has(propName, updatedValue).vertices()));
         graph.tx().rollback();
     }
 
+    /**
+     * Check whether given iterable does not contain any valid vertex
+     * This function returns true if either of the following conditions holds:
+     * 1) the iterable is empty
+     * 2) all vertices in the iterable are phantom vertices
+     * @param vertices An iterable of vertices
+     * @return boolean indicating if given vertices do not exist
+     */
+    private boolean verticesRemoved(Iterable<JanusGraphVertex> vertices) {
+        if (Iterables.isEmpty(vertices)) {
+            return true;
+        }
+        StandardJanusGraphTx queryTx = (StandardJanusGraphTx) graph.newTransaction();
+        for (JanusGraphVertex v : vertices) {
+            if (!graph.edgeQuery(v.longId(), graph.vertexExistenceQuery, queryTx.getTxHandle()).isEmpty()) {
```

Comment:
What if there are valid non-ghost vertices within the `Iterable` that just have no neighbors?
I think they are considered "removed" here, right? Should they?

Reply:
I guess you are misguided by the method name, "edgeQuery". It does not check neighbors, the method name is "edgeQuery" because almost everything is an edge in JanusGraph to my understanding. The method queries "edgeStore", which stores all vertices, properties, and edges.

Reply:
Thanks for pointing that out! You are completely right, I misunderstood `edgeQuery`.